kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/PartitionPruningSuite.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults
+
+import scala.collection.JavaConverters._
+
+import io.delta.golden.GoldenTableUtils.goldenTablePath
+import io.delta.kernel.defaults.utils.{TestRow, TestUtils}
+import io.delta.kernel.expressions.{Column, Expression, Predicate}
+import io.delta.kernel.expressions.Literal.ofInt
+import org.scalatest.funsuite.AnyFunSuite
+
+class PartitionPruningSuite extends AnyFunSuite with TestUtils {
+  test("partition pruning: simple filter") {
```

Comment:
actually, can you test with multiple levels of partitions?

I just tested this locally with 2 levels of partitions (part_a, part_b) and it failed
```
[id: bigint, part_a: bigint, part_b: bigint, col_1: bigint, col_2: bigint, col_3: bigint, col_4: bigint, col_5: bigint, col_6: bigint, col_7: bigint, col_8: bigint, col_9: bigint, col_10: bigint, col_11: bigint, col_12: bigint, col_13: bigint, col_14: bigint, col_15: bigint, col_16: bigint]
```

```
[info] PartitionPruningSuite:
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
class io.delta.kernel.internal.snapshot.SnapshotManager :: startVersion: 240, versionToLoad: Optional.empty
class io.delta.kernel.internal.snapshot.SnapshotManager :: startVersion: 240
class io.delta.kernel.internal.snapshot.SnapshotManager :: newFiles: [00000000000000000240.checkpoint.parquet, 00000000000000000240.json, 00000000000000000241.json, 00000000000000000242.json, 00000000000000000243.json, 00000000000000000244.json, 00000000000000000245.json, 00000000000000000246.json, 00000000000000000247.json, 00000000000000000248.json, 00000000000000000249.json]
class io.delta.kernel.internal.snapshot.SnapshotManager ::
checkpoints: [00000000000000000240.checkpoint.parquet]
deltas: [00000000000000000240.json, 00000000000000000241.json, 00000000000000000242.json, 00000000000000000243.json, 00000000000000000244.json, 00000000000000000245.json, 00000000000000000246.json, 00000000000000000247.json, 00000000000000000248.json, 00000000000000000249.json]
class io.delta.kernel.internal.snapshot.SnapshotManager :: lastCheckpoint: CheckpointInstance{version=9223372036854775807, numParts=Optional.empty}
class io.delta.kernel.internal.snapshot.SnapshotManager :: checkpointFiles: [CheckpointInstance{version=240, numParts=Optional.empty}]
class io.delta.kernel.internal.snapshot.SnapshotManager :: newCheckpointOpt: Optional[CheckpointInstance{version=240, numParts=Optional.empty}]
class io.delta.kernel.internal.snapshot.SnapshotManager :: newCheckpointVersion: 240
class io.delta.kernel.internal.snapshot.SnapshotManager :: deltasAfterCheckpoint: [00000000000000000241.json, 00000000000000000242.json, 00000000000000000243.json, 00000000000000000244.json, 00000000000000000245.json, 00000000000000000246.json, 00000000000000000247.json, 00000000000000000248.json, 00000000000000000249.json]
class io.delta.kernel.internal.snapshot.SnapshotManager :: deltaVersions: [241, 242, 243, 244, 245, 246, 247, 248, 249]
class io.delta.kernel.internal.snapshot.SnapshotManager :: Loading version 249 starting from checkpoint version 240.
[info] - foo *** FAILED ***
[info]   java.lang.UnsupportedOperationException: Child vectors are not available.
[info]   at io.delta.kernel.data.ColumnVector.getChild(ColumnVector.java:208)
[info]   at io.delta.kernel.defaults.internal.data.DefaultRowBasedColumnarBatch$DefaultColumnVector.getChild(DefaultRowBasedColumnarBatch.java:232)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.visitColumn(DefaultExpressionEvaluator.java:371)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.visitColumn(DefaultExpressionEvaluator.java:249)
[info]   at io.delta.kernel.defaults.internal.expressions.ExpressionVisitor.visit(ExpressionVisitor.java:62)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.visitElementAt(DefaultExpressionEvaluator.java:392)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.visitElementAt(DefaultExpressionEvaluator.java:249)
[info]   at io.delta.kernel.defaults.internal.expressions.ExpressionVisitor.visitScalarExpression(ExpressionVisitor.java:91)
[info]   at io.delta.kernel.defaults.internal.expressions.ExpressionVisitor.visit(ExpressionVisitor.java:58)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.visitPartitionValue(DefaultExpressionEvaluator.java:386)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.visitPartitionValue(DefaultExpressionEvaluator.java:249)
[info]   at io.delta.kernel.defaults.internal.expressions.ExpressionVisitor.visit(ExpressionVisitor.java:56)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.evalBinaryExpressionChildren(DefaultExpressionEvaluator.java:405)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.visitComparator(DefaultExpressionEvaluator.java:294)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.visitComparator(DefaultExpressionEvaluator.java:249)
[info]   at io.delta.kernel.defaults.internal.expressions.ExpressionVisitor.visitScalarExpression(ExpressionVisitor.java:89)
[info]   at io.delta.kernel.defaults.internal.expressions.ExpressionVisitor.visit(ExpressionVisitor.java:58)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.evalBinaryExpressionChildren(DefaultExpressionEvaluator.java:406)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.visitAnd(DefaultExpressionEvaluator.java:258)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator$ExpressionEvalVisitor.visitAnd(DefaultExpressionEvaluator.java:249)
[info]   at io.delta.kernel.defaults.internal.expressions.ExpressionVisitor.visitScalarExpression(ExpressionVisitor.java:80)
[info]   at io.delta.kernel.defaults.internal.expressions.ExpressionVisitor.visit(ExpressionVisitor.java:58)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultExpressionEvaluator.eval(DefaultExpressionEvaluator.java:66)
[info]   at io.delta.kernel.defaults.internal.expressions.DefaultPredicateEvaluator.eval(DefaultPredicateEvaluator.java:68)
[info]   at io.delta.kernel.internal.ScanImpl.lambda$applyPartitionPruning$8(ScanImpl.java:180)
[info]   at io.delta.kernel.utils.CloseableIterator$1.next(CloseableIterator.java:55)
[info]   at io.delta.kernel.defaults.utils.TestUtils$CloseableIteratorOps.forEach(TestUtils.scala:46)
[info]   at io.delta.kernel.defaults.utils.TestUtils.readSnapshot(TestUtils.scala:104)
[info]   at io.delta.kernel.defaults.utils.TestUtils.readSnapshot$(TestUtils.scala:80)
[info]   at io.delta.kernel.defaults.PartitionPruningSuite.readSnapshot(PartitionPruningSuite.scala:26)
[info]   at io.delta.kernel.defaults.utils.TestUtils.checkTable(TestUtils.scala:216)
[info]   at io.delta.kernel.defaults.utils.TestUtils.checkTable$(TestUtils.scala:186)
[info]   at io.delta.kernel.defaults.PartitionPruningSuite.checkTable(PartitionPruningSuite.scala:26)
[info]   at io.delta.kernel.defaults.PartitionPruningSuite.$anonfun$new$3(PartitionPruningSuite.scala:47)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
```

Reply:
```scala
  test("foo") {
    val filter = predicate("=", col("part_a"), ofInt(999))

    checkTable(
      path = "/Users/scott.sandre/tmp_delta_tables/table_05__1_000_000__complex_schema",
      expectedAnswer = Nil, // i just want to time how long it takes to read all the data
      readCols = null, 
      filter = filter
    )
  }
```

Reply:
If you care, i updated the temporary test to just get all of the files in the table, but it had the same error

```scala
  def getFiles(
      snapshot: Snapshot,
      filter: Predicate = null,
      tableClient: TableClient = defaultTableClient): Seq[FileStatus] = {
    var scanBuilder = snapshot.getScanBuilder(tableClient)

    if (filter != null) {
      scanBuilder = scanBuilder.withFilter(tableClient, filter)
    }

    val scan = scanBuilder.build()

    val fileIter = scan.getScanFiles(tableClient)

    val output = new ArrayBuffer[FileStatus]()
    fileIter.forEach { fileColumnarBatch =>
      fileColumnarBatch.getRows().forEach { scanFileRow =>
        output += InternalScanFileUtils.getAddFileStatus(scanFileRow)
      }
    }
    output
  }
```

Reply:
All levels work. I think the PR when you looked at has one commit missing: 9fb71285647c06d8fa2861445d7e66e690fe1f63